### PR TITLE
Fix for Pumps Turning Off and an Overhaul of the C++ Code in Pumps

### LINKED ISSUE #9427: Zero heating consumption when forcing pump flow rate and no warning or error
Issue overview
--------------

In the process of investigating another issue, I discovered something strange. I set up a model that has a PlantLoop with a DistrictHeating and a Pump:VariableSpeed object and serves ZoneHVAC:Baseboard:Convective:Water object(s).

* **noforce_pump**: I run the first simulation, with the Pump:Variable on autosized `Design Maximum Flow Rate {m3/s}` and  no `Design Minimum Flow Rate {m3/s}`
* **force_pump**: In the second simulation I retrieve the Pump Design Flow Rate from the first run (noforce_pump), and harcode that into both fields

If I do that with 1 Zone (And 1 ZoneHVAC:Baseboard:Convective:Water object), it goes well in both case.
If I use 2 zones, the force_pump has zero heating consumption. No Warning or errors

image

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): Ubuntu 20.04
 - Version of EnergyPlus (if using an intermediate build, include SHA): V22-1-0 (but 9.6.0 too)

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file here): https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/9000s/9427
 - [x] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by RKStrand**
Pull request overview
---------------------
 - Fixes #9427 
 - It was noticed in a particular file that variable speed pumps were totally turning off and this led to no heating for a case when the pump maximum and minimum flow rate were specified by the user.  In the same file, a switch to autosizing those two parameters showed the pumps operating as expected.  The problem was that in the hard-wired input file both the maximum and minimum were exactly the same.  This case was not trapped and resulted in no pump operation.  However, if a user had chosen any other minimum flow rate the file would run with pumps operating and heating reported.  This has now been fixed.  In addition, it was requested that the pump module be significantly cleaned up and this was achieved.  The pump module is much cleaner now.  No real differences in the output files other than a minor difference in the error file due to modifications made to simplify the error messages that are produced.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Review by amirroth — APPROVED** [review]
I won't comment on the underlying issue that precipitated this branch and I've made plenty of comments during the restructuring already. It's always possible to do more (spoiler: I am seeing some opportunities now as I write this) but as some point you just have to stop. The usage of old idioms has been reduced as much as it has in any other module that has undergone this sort of pass. I am satisfied. Thanks @RKStrand!

File: tst/EnergyPlus/unit/Pumps.unit.cc [hunk 1]
```diff
@@ -650,9 +650,9 @@ TEST_F(EnergyPlusFixture, VariableSpeedPump_MinFlowGreaterThanMax)
     Pumps::GetPumpInput(*state);
 
     std::string const error_string = delimited_string({
-        "   ** Warning ** GetPumpInput: Pump:VariableSpeed=\"SUPPLY INLET PUMP\", Invalid 'Design Minimum Flow Rate'",
-        "   **   ~~~   ** Entered Value=[2.00000E-003] is above the Design Maximum Flow Rate=[1.00000E-003].",
-        "   **   ~~~   ** Reseting value of 'Design Minimum Flow Rate' to the value of 'Design Maximum Flow Rate'.",
+        "   ** Warning ** GetPumpInput:  Pump:VariableSpeed = SUPPLY INLET PUMP, Invalid 'Design Minimum Flow Rate'",
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/Pumps.unit.cc]: Why this change of punctuation? double space at last (the quotes were also clearer but that's just my two cents)

File: src/EnergyPlus/Pumps.hh [hunk 2]
```diff
@@ -122,131 +113,93 @@ namespace Pumps {
...
-        }
+        int ManualRPMSchedIndex = 0;
+        int LowerPsetSchedIndex = 0;
+        int UpperPsetSchedIndex = 0;
+        int MinRPMSchedIndex = 0;
+        int MaxRPMSchedIndex = 0;
+        ControlTypeVFD VFDControlType = ControlTypeVFD::Invalid; // VFDControlType
+        Real64 MaxRPM = 0.0;                                     // Maximum RPM range value - schedule limit
+        Real64 MinRPM = 0.0;                                     // Minimum RPM range value - schedule limit
+        Real64 PumpActualRPM = 0.0;                              // RPM recalculated from final flow through the loop
```
Comment (jmarrec) [inline src/EnergyPlus/Pumps.hh:124]: I definitely like the changes in the header. Member init at declaration is much clearer to me.

I wonder how often the Schedule Names are retrieved and used and whether caching (the `std::string xxxSchedName`) was a good idea to begin with. I do not know if this has been discussed / ruled upon before either though.

File: src/EnergyPlus/Pumps.cc [hunk 3]
```diff
@@ -105,19 +105,8 @@ using DataHVACGlobals::ForceOff;
...
-
-std::map<PumpType, std::string> cPumpTypes = {
-    {PumpType::VarSpeed, cPump_VarSpeed},
-    {PumpType::ConSpeed, cPump_ConSpeed},
-    {PumpType::Cond, cPump_Cond},
-    {PumpType::Bank_VarSpeed, cPumpBank_VarSpeed},
-    {PumpType::Bank_ConSpeed, cPumpBank_ConSpeed},
-};
+constexpr std::array<std::string_view, static_cast<int>(PumpType::Num)> pumpTypeIDFNames = {
+    "Pump:VariableSpeed", "Pump:ConstantSpeed", "Pump:VariableSpeed:Condensate", "HeaderedPumps:VariableSpeed", "HeaderedPumps:ConstantSpeed"};
```
Comment (jmarrec) [inline src/EnergyPlus/Pumps.cc]: Won't make a difference on gcc, but I think this should be `static constexpr`

File: src/EnergyPlus/Pumps.cc [hunk 4]
```diff
@@ -239,6 +228,10 @@ void GetPumpInput(EnergyPlusData &state)
     Real64 constexpr StartTemp(100.0); // Standard Temperature across code to calculated Steam density
     static constexpr std::string_view RoutineName("GetPumpInput: ");
     static constexpr std::string_view RoutineNameNoColon("GetPumpInput");
+    constexpr std::array<std::string_view, static_cast<int>(PumpControlType::Num)> pumpCtrlTypeNamesUC{"CONTINUOUS", "INTERMITTENT"};
+    constexpr std::array<std::string_view, static_cast<int>(ControlTypeVFD::Num)> controlTypeVFDNamesUC{"MANUALCONTROL", "PRESSURESETPOINTCONTROL"};
+    constexpr std::array<std::string_view, static_cast<int>(PowerSizingMethod::Num)> powerSizingMethodNamesUC{"POWERPERFLOW",
+                                                                                                              "POWERPERFLOWPERPRESSURE"};
```
Comment (jmarrec) [inline src/EnergyPlus/Pumps.cc]:
```suggestion
    static constexpr std::array<std::string_view, static_cast<int>(PumpControlType::Num)> pumpCtrlTypeNamesUC{"CONTINUOUS", "INTERMITTENT"};
    static constexpr std::array<std::string_view, static_cast<int>(ControlTypeVFD::Num)> controlTypeVFDNamesUC{"MANUALCONTROL", "PRESSURESETPOINTCONTROL"};
    static constexpr std::array<std::string_view, static_cast<int>(PowerSizingMethod::Num)> powerSizingMethodNamesUC{"POWERPERFLOW",
                                                                                                              "POWERPERFLOWPERPRESSURE"};
```

File: src/EnergyPlus/Pumps.cc (hunk 3 above)
Comment (jmarrec) [inline src/EnergyPlus/Pumps.cc]:
```suggestion
static constexpr std::array<std::string_view, static_cast<int>(PumpType::Num)> pumpTypeIDFNames = {
    "Pump:VariableSpeed", "Pump:ConstantSpeed", "Pump:VariableSpeed:Condensate", "HeaderedPumps:VariableSpeed", "HeaderedPumps:ConstantSpeed"};
```

File: src/EnergyPlus/Pumps.cc [hunk 5]
```diff
@@ -251,25 +244,23 @@ void GetPumpInput(EnergyPlusData &state)
     int NumVarSpeedPumps;
     int NumConstSpeedPumps;
     int NumCondensatePumps;
```
Comment (jmarrec) [inline src/EnergyPlus/Pumps.cc]: Not saying we need to do it, but IMHO declaration of variables at top of routine is a code smell.

*Uninitialized* declarations is even worse.

File: src/EnergyPlus/Pumps.cc [hunk 6]
```diff
@@ -281,950 +272,898 @@ void GetPumpInput(EnergyPlusData &state)
...
+        auto &thisPump = state.dataPumps->PumpEquip(PumpNum);
         state.dataInputProcessing->inputProcessor->getObjectItem(state,
                                                                  cCurrentModuleObject,
-                                                                 NumVarPump,
-                                                                 state.dataIPShortCut->cAlphaArgs,
+                                                                 PumpNum,
+                                                                 thisInput->cAlphaArgs,
                                                                  NumAlphas,
-                                                                 state.dataIPShortCut->rNumericArgs,
+                                                                 thisInput->rNumericArgs,
```
Comment (jmarrec) [inline src/EnergyPlus/Pumps.cc:287]: (Are we eventually moving to the epJSON stuff? eg stuff like here: https://github.com/NREL/EnergyPlus/blob/7401e1fee9615a0fc571718d1845c732182b8529/src/EnergyPlus/SystemAvailabilityManager.cc#L1302-L1343)

File: src/EnergyPlus/Pumps.cc [hunk 7]
```diff
@@ -281,950 +272,898 @@ void GetPumpInput(EnergyPlusData &state)
...
     state.dataPumps->PumpEquipReport.allocate(state.dataPumps->NumPumps);
     auto &cCurrentModuleObject = state.dataIPShortCut->cCurrentModuleObject;
-    cCurrentModuleObject = cPump_VarSpeed;
+    cCurrentModuleObject = pumpTypeIDFNames[static_cast<int>(PumpType::VarSpeed)];
+    auto &thisInput = state.dataIPShortCut;
 
-    for (NumVarPump = 1; NumVarPump <= NumVarSpeedPumps; ++NumVarPump) {
-        PumpNum = NumVarPump;
+    for (PumpNum = 1; PumpNum <= NumVarSpeedPumps; ++PumpNum) {
+        auto &thisPump = state.dataPumps->PumpEquip(PumpNum);
```
Comment (jmarrec) [inline src/EnergyPlus/Pumps.cc:281]: :+1: clears out so much text below...

File: src/EnergyPlus/Pumps.cc [hunk 8]
```diff
@@ -281,950 +272,898 @@ void GetPumpInput(EnergyPlusData &state)
...
+            if (thisPump.ZoneNum > 0) {
+                thisPump.HeatLossesToZone = true;
+                if (!thisInput->lNumericFieldBlanks(7)) {
+                    thisPump.SkinLossRadFraction = thisInput->rNumericArgs(7);
                 }
             } else {
                 ShowSevereError(state,
-                                cCurrentModuleObject + "=\"" + state.dataIPShortCut->cAlphaArgs(1) + "\" invalid " +
-                                    state.dataIPShortCut->cAlphaFieldNames(7) + "=\"" + state.dataIPShortCut->cAlphaArgs(7) + "\" not found.");
+                                format("{}={} invalid {}={} not found.",
```
Comment (jmarrec) [inline src/EnergyPlus/Pumps.cc]: Is the removal of the `"` inside the format string wanted?

File: src/EnergyPlus/Pumps.cc [hunk 9]
```diff
@@ -281,950 +272,898 @@ void GetPumpInput(EnergyPlusData &state)
...
         }
-        if (!state.dataIPShortCut->lAlphaFieldBlanks(8)) {
-            if (state.dataIPShortCut->cAlphaArgs(8) == "POWERPERFLOW") {
-                state.dataPumps->PumpEquip(PumpNum).powerSizingMethod = PowerSizingMethod::SizePowerPerFlow;
-            } else if (state.dataIPShortCut->cAlphaArgs(8) == "POWERPERFLOWPERPRESSURE") {
-                state.dataPumps->PumpEquip(PumpNum).powerSizingMethod = PowerSizingMethod::SizePowerPerFlowPerPressure;
-            } else {
+        if (!thisInput->lAlphaFieldBlanks(8)) {
+            thisPump.powerSizingMethod = static_cast<PowerSizingMethod>(
+                getEnumerationValue(powerSizingMethodNamesUC, UtilityRoutines::MakeUPPERCase(state.dataIPShortCut->cAlphaArgs(8))));
```
Comment (jmarrec) [inline src/EnergyPlus/Pumps.cc:1082]: I've seen this idiom a lot until now... Should we make a higher level helper? it'll save one line, but may be clearer

```
template <typename T>
constexpr T getEnumFromString(const gsl::span<const std::string_view> sList,
                              const std::string_view s) {
  return static_cast<T>(getEnumerationValue(sList, s));
}
```
Usage : `thisPump.powerSizingMethod = getEnumFromString<PowerSizingMethod>(powerSizingMethodNamesUC, state.dataIPShortCut->cAlphaArgs(8));`

(not sure whether the MakeUPPERCase is needed nowadays or not, but that can be added)

File: src/EnergyPlus/Pumps.cc [hunk 10]
```diff
@@ -1902,56 +1800,55 @@ void CalcPumps(EnergyPlusData &state, int const PumpNum, Real64 const FlowReques
...
 
     //  ! If this is a variable speed pump
-    if (BITF_TEST_ANY(BITF(state.dataPumps->PumpEquip(PumpNum).pumpType),
-                      BITF(PumpType::VarSpeed) | BITF(PumpType::Bank_VarSpeed) | BITF(PumpType::Cond))) {
-        if (DataPlant::CompData::getPlantComponent(state, state.dataPumps->PumpEquip(PumpNum).plantLoc).FlowCtrl ==
-            DataBranchAirLoopPlant::ControlType::SeriesActive) {
-            state.dataPumps->PumpMassFlowRate = 0.0;
+    if (thisPump.pumpType == PumpType::VarSpeed || thisPump.pumpType == PumpType::Bank_VarSpeed || thisPump.pumpType == PumpType::Cond) {
+        if (DataPlant::CompData::getPlantComponent(state, thisPump.plantLoc).FlowCtrl == DataBranchAirLoopPlant::ControlType::SeriesActive) {
+            daPumps->PumpMassFlowRate = 0.0;
```
Comment (jmarrec) [inline src/EnergyPlus/Pumps.cc:1827]: I'm glad to see that BITF stuff gone. 👍

File: src/EnergyPlus/Pumps.cc [hunk 11]
```diff
@@ -1233,169 +1172,155 @@ void GetPumpInput(EnergyPlusData &state)
...
-                                state.dataPumps->PumpEquipReport(PumpNum).NumPumpsOperating,
+                                thisPumpRep.NumPumpsOperating,
                                 OutputProcessor::SOVTimeStepType::System,
                                 OutputProcessor::SOVStoreType::Average,
-                                state.dataPumps->PumpEquip(PumpNum).Name);
+                                thisPump.Name);
         } break;
         default: {
-            ShowFatalError(state, format("Invalid Pump Type = {}", state.dataPumps->PumpEquip(PumpNum).pumpType));
+            assert(false);
```
Comment (jmarrec) [inline src/EnergyPlus/Pumps.cc:1246]: missing `#include <cassert>` at beginning of file.

File: src/EnergyPlus/Pumps.cc [hunk 12]
```diff
@@ -2366,59 +2208,49 @@ void ReportPumps(EnergyPlusData &state, int const PumpNum)
...
+    auto &thisPumpRep = state.dataPumps->PumpEquipReport(PumpNum);
+
+    PumpType = thisPump.pumpType;
+    OutletNode = thisPump.OutletNodeNum;
+    auto &thisOutNode = state.dataLoopNodes->Node(OutletNode);
+    auto &daPumps = state.dataPumps;
+
+    if (daPumps->PumpMassFlowRate <= DataBranchAirLoopPlant::MassFlowTolerance) {
+        new (&(state.dataPumps->PumpEquipReport(PumpNum))) ReportVars();
+        thisPumpRep.OutletTemp = thisOutNode.Temp;
```
Comment (jmarrec) [inline src/EnergyPlus/Pumps.cc:2236]: I'm confused by this one (as in I'm not sure what that construct does). Could you explain the `new` line please? I'm getting lost.

**Review by jmarrec — COMMENTED** [review]
Left a couple of small comments. This looks much better!

File: tst/EnergyPlus/unit/Pumps.unit.cc (hunk 1 above)
Comment (RKStrand) [reply tst/EnergyPlus/unit/Pumps.unit.cc]: I personally didn't feel like the quotes added much.  The double space between the GetPumpInput: and Pump:VariableSpeed...that can be gotten rid of.  But I'm also just one opinion.  Doesn't matter to me.  Whatever the team wants I can implement.

File: src/EnergyPlus/Pumps.hh (hunk 2 above)
Comment (RKStrand) [reply src/EnergyPlus/Pumps.hh:124]: @amirroth requested this and I ran with it.  The only benefit I see with keeping the schedule name is for debugging, but that's of limited value I think.

Comment (amirroth) [reply src/EnergyPlus/Pumps.hh:124]: No, caching schedule names was not a good idea. It has not been discussed in my recollection. PS There are multiple other instances of this type of thing. Too much string copying in general in EnergyPlus.

File: src/EnergyPlus/Pumps.cc (hunk 3 above)
Comment (RKStrand) [reply src/EnergyPlus/Pumps.cc]: I went with @amirroth's suggestion but if this works and is better, that's fine with me.

File: src/EnergyPlus/Pumps.cc (hunk 8 above)
Comment (RKStrand) [reply src/EnergyPlus/Pumps.cc]: Again, a matter of preference.  I can put them back in if that makes more sense to people.

File: src/EnergyPlus/Pumps.cc (hunk 3 above)
Comment (amirroth) [reply src/EnergyPlus/Pumps.cc]: Hmmm. Good to know. I only use gcc/g++.

File: src/EnergyPlus/Pumps.cc (hunk 9 above)
Comment (RKStrand) [reply src/EnergyPlus/Pumps.cc:1082]: Good idea--I'd ask that we pin that and come back later.  There are lots of places in the code where this is going to be an issue.  Seems like it would be best if someone went through the code and dealt with these all at once.

Comment (amirroth) [reply src/EnergyPlus/Pumps.cc:1082]: Sure, why not.

File: src/EnergyPlus/Pumps.cc (hunk 11 above)
Comment (RKStrand) [reply src/EnergyPlus/Pumps.cc:1246]: I'll add it.

File: src/EnergyPlus/Pumps.cc (hunk 12 above)
Comment (RKStrand) [reply src/EnergyPlus/Pumps.cc:2236]: I think this basically resets the variable/structure back to its default value.  In this case, it zeroes everything out.  The code it replace did that with one exception (OutletTemp)., hence the next line.

Comment (amirroth) [reply src/EnergyPlus/Pumps.cc:2236]: This "new with placement". It's been a C++ feature since the beginning. By giving an address argument to `new`, you suppress the memory allocation and call the constructor (in this case `ReportVars()`) on the address provided. 

It's similar to `state.dataPumps->PumpEquipReport(PumpNum) = ReportVars();` but it doesn't construct and copy, it just constructs into the destination. Yes, I know about RVO. But why rely on RVO when you have this?

File: src/EnergyPlus/Pumps.cc (hunk 5 above)
Comment (RKStrand) [reply src/EnergyPlus/Pumps.cc]: Ok, they will be initialized.  At least it won't be "even worse" now.

File: src/EnergyPlus/Pumps.cc (hunk 6 above)
Comment (RKStrand) [reply src/EnergyPlus/Pumps.cc:287]: I don't know.  Perhaps @amirroth can comment.  A lot of work has been done here already.  I don't know the plans long term for epJSON in EnergyPlus.

Comment (amirroth) [reply src/EnergyPlus/Pumps.cc:287]: @jmarrec We are moving towards epJSON based inputs, but we are also potentially moving to a system in which some of that code will be automatically generated from the epJSON schema (a la OpenStudio). It doesn't make sense to continue doing this transition manually until we see what we can do automatically and what that actually looks like. And as @RKStrand mentioned, this was already a lot of surgery for a first time doing a pass like this.

**Comment by RKStrand** [comment]
@jmarrec I have hopefully corrected all of the issues here other than the one I asked that we deal with later throughout the entire code.  These should all be in the commit that I made a little while ago.  If I missed anything, please let me know and I will correct that.

File: src/EnergyPlus/Pumps.cc (hunk 3 above)
Comment (jmarrec) [reply src/EnergyPlus/Pumps.cc]: @lefticus made a c++ weekly on it the topic recently.

https://youtu.be/4pKtPWcl1Go

(I'm not saying he said to do this specific thing, this was my opinion. I tested gcc and compiler explorer and didn't made a difference in this specific case at -O3)

File: src/EnergyPlus/Pumps.hh (hunk 2 above)
Comment (jmarrec) [reply src/EnergyPlus/Pumps.hh:124]: Thanks for the replies! (I view this as an opportunity to learn)

I agree there's too much caching generally.

For debugging you should be able to do state->Xxxx anyways so that's not a concern.

File: src/EnergyPlus/Pumps.cc (hunk 6 above)
Comment (jmarrec) [reply src/EnergyPlus/Pumps.cc:287]: Totally agree this is completely out of scope. I put it inside parentheses but should probably have added a "Side note" disclaimer too :)

File: src/EnergyPlus/Pumps.cc (hunk 12 above)
Comment (jmarrec) [reply src/EnergyPlus/Pumps.cc:2236]: Today I learned! Thanks

File: src/EnergyPlus/Pumps.cc (hunk 3 above)
Comment (amirroth) [reply src/EnergyPlus/Pumps.cc]: I saw that too. I think it applies to header files only.

File: doc/input-output-reference/src/overview/group-pumps.tex [hunk 13]
```diff
@@ -64,7 +64,7 @@ \subsubsection{Inputs}\label{inputs-037}
 
 \paragraph{Field: Design Minimum Flow Rate}\label{field-design-minimum-flow-rate}
 
-This field contains the minimum volume flow rate while operating in variable flow capacity rate in cubic meters per second. This field is autosizable. When set to autosize the value in the input field Design Minimum Flow Rate Sizing Factor is used to scale the minimum flow rate as a fraction of the Design Maximum Flow Rate.
+This field contains the minimum volume flow rate while operating in variable flow capacity rate in cubic meters per second. This field is autosizable. When set to autosize the value in the input field Design Minimum Flow Rate Sizing Factor is used to scale the minimum flow rate as a fraction of the Design Maximum Flow Rate.  Note that this Design Minimum Flow Rate should be less than the Design Maximum Flow Rate for a variable speed pump.  If the user enters a Design Minimum Flow Rate that is more than 99\% of the Design Maximum Flow Rate, a warning message is produced and the Design Minimum Flow Rate is reset to 99\% of the Design Maximum Flow Rate.
```
Comment (Myoldmopar) [inline doc/input-output-reference/src/overview/group-pumps.tex:67]: OK

File: src/EnergyPlus/Pumps.hh [hunk 14]
```diff
@@ -48,9 +48,6 @@
 #ifndef Pumps_hh_INCLUDED
 #define Pumps_hh_INCLUDED
 
-// ObjexxFCL Headers
-#include <ObjexxFCL/Array1D.hh>
```
Comment (Myoldmopar) [inline src/EnergyPlus/Pumps.hh:52]: Nice.

File: src/EnergyPlus/Pumps.hh [hunk 15]
```diff
@@ -93,12 +90,6 @@ namespace Pumps {
         Num
     };
 
-    extern std::string const cPump_VarSpeed;
-    extern std::string const cPump_ConSpeed;
-    extern std::string const cPump_Cond;
-    extern std::string const cPumpBank_VarSpeed;
-    extern std::string const cPumpBank_ConSpeed;
```
Comment (Myoldmopar) [inline src/EnergyPlus/Pumps.hh:100]: Yes.

File: src/EnergyPlus/Pumps.hh [hunk 16]
```diff
@@ -122,131 +113,93 @@ namespace Pumps {
...
-        bool OneTimePressureWarning;
-        bool HeatLossesToZone;        // if true then pump losses added to surrounding zone
-        int ZoneNum;                  // index for zone surrounding pump
-        Real64 SkinLossRadFraction;   // radiative split for skin losses to zone
-        bool LoopSolverOverwriteFlag; // loop solver overwrite for determining pump minimum flow rate
+        bool OneTimePressureWarning = true;
+        bool HeatLossesToZone = false;        // if true then pump losses added to surrounding zone
+        int ZoneNum = 0;                      // index for zone surrounding pump
+        Real64 SkinLossRadFraction = 0.0;     // radiative split for skin losses to zone
+        bool LoopSolverOverwriteFlag = false; // loop solver overwrite for determining pump minimum flow rate
```
Comment (Myoldmopar) [inline src/EnergyPlus/Pumps.hh:186]: Much cleaner

File: src/EnergyPlus/Pumps.hh [hunk 17]
```diff
@@ -122,131 +113,93 @@ namespace Pumps {
...
+        int NumPumpsOperating = 0;          // Used in pump bank. reports how many pumps are ON
+        Real64 PumpMassFlowRate = 0.0;      // Mass flow rate of the pump
+        Real64 PumpHeattoFluid = 0.0;       // Heat transfer from pump to fluid (W)
+        Real64 PumpHeattoFluidEnergy = 0.0; // Pump Energy dissipated into fluid stream
+        Real64 OutletTemp = 0.0;            // Pump outlet temperature
+        Real64 ShaftPower = 0.0;            // Power input at the shaft
+        Real64 ZoneTotalGainRate = 0.0;     // total pump skin losses to zone (W)
+        Real64 ZoneTotalGainEnergy = 0.0;   // total pump skin losses to zone energy (J)
+        Real64 ZoneConvGainRate = 0.0;      // pump skin losses convecting to zone air (W)
+        Real64 ZoneRadGainRate = 0.0;       // pump skin losses radiating to inside of zone (W)
```
Comment (Myoldmopar) [inline src/EnergyPlus/Pumps.hh:202]: One possible addition is to just eliminate this extra report vars struct entirely.  The variables can live on the base pump data structure, as they are both allocated to the same size every single time anyway.  No big deal, just noting for the future.

File: src/EnergyPlus/Pumps.hh [hunk 18]
```diff
@@ -292,8 +245,8 @@ struct PumpsData : BaseGlobalStruct
     Real64 PumpHeattoFluid = 0.0;  // Pump Power dissipated in fluid stream
     Real64 Power = 0.0;            // Pump Electric power
     Real64 ShaftPower = 0.0;       // Power passing through pump shaft
```
Comment (Myoldmopar) [inline src/EnergyPlus/Pumps.hh:247]: This data structure needs to be cleaned out.  There is no reason inside EnergyPlus that we need a global-ish variable to hold `Power`.  Each pump can hold its own value of power consumption.  Whenever this change is made, it needs to be made cautiously because it can easily lead to diffs.

File: tst/EnergyPlus/unit/Pumps.unit.cc [hunk 19]
```diff
@@ -651,8 +651,8 @@ TEST_F(EnergyPlusFixture, VariableSpeedPump_MinFlowGreaterThanMax)
 
     std::string const error_string = delimited_string({
         "   ** Warning ** GetPumpInput: Pump:VariableSpeed=\"SUPPLY INLET PUMP\", Invalid 'Design Minimum Flow Rate'",
-        "   **   ~~~   ** Entered Value=[2.00000E-003] is above the Design Maximum Flow Rate=[1.00000E-003].",
-        "   **   ~~~   ** Reseting value of 'Design Minimum Flow Rate' to the value of 'Design Maximum Flow Rate'.",
+        "   **   ~~~   ** Entered Value=[2.00000E-003] is above or too close (equal) to the Design Maximum Flow Rate=[1.00000E-003].",
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/Pumps.unit.cc:654]: :+1:

File: tst/EnergyPlus/unit/Pumps.unit.cc [hunk 20]
```diff
@@ -661,4 +661,59 @@ TEST_F(EnergyPlusFixture, VariableSpeedPump_MinFlowGreaterThanMax)
...
+
+    EXPECT_TRUE(compare_err_stream(error_string, true));
+
+    // Should have reset the min to 99% of the the max, max not impacted
+    Real64 expectedAnswer = 0.99 * 0.001;
+    Real64 constexpr allowableTolerance = 0.00001;
+    EXPECT_NEAR(state->dataPumps->PumpEquip(1).MinVolFlowRate, expectedAnswer, allowableTolerance);
+    expectedAnswer = 0.001;
+    EXPECT_NEAR(state->dataPumps->PumpEquip(1).NomVolFlowRate, expectedAnswer, allowableTolerance);
+}
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/Pumps.unit.cc:717]: Yes, great unit test!

File: src/EnergyPlus/Pumps.cc [hunk 21]
```diff
@@ -105,19 +106,8 @@ using DataHVACGlobals::ForceOff;
...
-std::string const cPumpBank_ConSpeed("HeaderedPumps:ConstantSpeed");
-
-std::map<PumpType, std::string> cPumpTypes = {
-    {PumpType::VarSpeed, cPump_VarSpeed},
-    {PumpType::ConSpeed, cPump_ConSpeed},
-    {PumpType::Cond, cPump_Cond},
-    {PumpType::Bank_VarSpeed, cPumpBank_VarSpeed},
-    {PumpType::Bank_ConSpeed, cPumpBank_ConSpeed},
-};
+static constexpr std::array<std::string_view, static_cast<int>(PumpType::Num)> pumpTypeIDFNames = {
```
Comment (Myoldmopar) [inline src/EnergyPlus/Pumps.cc:109]: Yep.

File: src/EnergyPlus/Pumps.cc [hunk 22]
```diff
@@ -281,950 +274,907 @@ void GetPumpInput(EnergyPlusData &state)
...
+                                                   DataLoopNode::ConnectionObjectType::PumpVariableSpeed,
+                                                   thisPump.Name,
+                                                   DataLoopNode::NodeFluidType::Water,
+                                                   DataLoopNode::ConnectionType::Outlet,
+                                                   NodeInputManager::CompFluidStream::Primary,
+                                                   ObjectIsNotParent);
+        TestCompSet(state, cCurrentModuleObject, thisPump.Name, thisInput->cAlphaArgs(2), thisInput->cAlphaArgs(3), "Water Nodes");
+
+        thisPump.PumpControl = static_cast<PumpControlType>(
+            getEnumerationValue(pumpCtrlTypeNamesUC, UtilityRoutines::MakeUPPERCase(state.dataIPShortCut->cAlphaArgs(4))));
```
Comment (Myoldmopar) [inline src/EnergyPlus/Pumps.cc:323]: Nice.

File: src/EnergyPlus/Pumps.cc [hunk 23]
```diff
@@ -281,950 +274,907 @@ void GetPumpInput(EnergyPlusData &state)
...
+        if (!thisInput->cAlphaArgs(5).empty()) { // Initialized to zero, don't get a schedule for an empty
+            thisPump.PumpScheduleIndex = GetScheduleIndex(state, thisInput->cAlphaArgs(5));
+            if (thisPump.PumpScheduleIndex <= 0) {
+                ShowWarningError(state,
+                                 format("{}{}=\"{}\", Invalid {}", RoutineName, cCurrentModuleObject, thisPump.Name, thisInput->cAlphaFieldNames(5)));
+                ShowContinueError(state, format("Schedule named =[{}]. was not found and will not be used.", thisInput->cAlphaArgs(5)));
+            }
+        }
+
+        thisPump.NomVolFlowRate = thisInput->rNumericArgs(1);
```
Comment (Myoldmopar) [inline src/EnergyPlus/Pumps.cc:343]: Interesting, I hadn't thought of aliasing the DataIPShortcut.  I would suggest that in the future, instead of doing that, just remove the usage of DataIPShortcut.  The purpose of that module was to hold a set of global arrays that each developer could use without having to manage and resize arrays themselves.  In C++, we can just pass vectors in and out of the input processor routines without having to presize them.  Then you can just use local variables named rNumericArgs, etc.  And there won't be any need to alias anything.

File: src/EnergyPlus/Pumps.cc (hunk 9 above)
Comment (Myoldmopar) [reply src/EnergyPlus/Pumps.cc:1082]: Yeah, using the template function to static cast it for us is a great idea.

File: src/EnergyPlus/Pumps.cc [hunk 24]
```diff
@@ -281,950 +274,907 @@ void GetPumpInput(EnergyPlusData &state)
...
+        thisPump.MotorEffic = thisInput->rNumericArgs(4);
+        thisPump.FracMotorLossToFluid = thisInput->rNumericArgs(5);
+        thisPump.PartLoadCoef[0] = thisInput->rNumericArgs(6);
+        thisPump.PartLoadCoef[1] = thisInput->rNumericArgs(7);
+        thisPump.PartLoadCoef[2] = thisInput->rNumericArgs(8);
+        thisPump.PartLoadCoef[3] = thisInput->rNumericArgs(9);
+        thisPump.MinVolFlowRate = thisInput->rNumericArgs(10);
+        if (thisPump.MinVolFlowRate == AutoSize) {
+            thisPump.minVolFlowRateWasAutosized = true;
+        } else if (!thisPump.NomVolFlowRateWasAutoSized && (thisPump.MinVolFlowRate > (minToMaxRatioMax * thisPump.NomVolFlowRate))) {
```
Comment (Myoldmopar) [inline src/EnergyPlus/Pumps.cc:361]: New condition here, got it.

**Review by Myoldmopar — APPROVED** [review]
Good to go.

**Comment by Myoldmopar** [comment]
This is good to go in, CI is good and everything is happy locally as well.
